app/models/registration.rb

```diff
@@ -4,13 +4,70 @@ class Registration < ApplicationRecord
   # Professors have extra privileges over assistants.
   enum role: [:student, :grader, :assistant, :professor]
 
+  validates_presence_of :course_id, :user_id
+  validates_inclusion_of :show_in_lists, in: [false, true]
+  
   # A registration is join model between a user and a course.
   belongs_to :user
   belongs_to :course
 
   has_one :term, through: :course
-  belongs_to :section, :primary_key => "crn"
+  has_many :registration_sections, dependent: :destroy
+  has_many :sections, through: :registration_sections
 
+  Section::types.each do |t, value|
+    # This reader is for convenience
+    define_method "#{t}_sections".to_sym do
+      self.sections.where(type: value)
+    end
+  end
+
+  attr_accessor :username
+  attr_accessor :orig_sections
```

Comment:
Assuming you have has_many :through declared correctly, setting the sections attribute to be the correct list of sections before save should update the join table correctly.

I'll go ahead and pull this now - the atomicity failure of a student having the wrong sections is correctable - but it's worth taking another look at this the next time you touch this code.